api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto

```diff
@@ -196,6 +197,27 @@ message ExternalProcessor {
   // will be added to StreamInfo's filter state under the namespace corresponding to the
   // ext_proc filter name.
   google.protobuf.Struct filter_metadata = 13;
+
+  // [#not-implemented-hide:]
+  // Set the response header size and number limits.
+  // When the ext_proc filter receives a response that exceeds these limits, it will ignore
+  // the response and and consider it as spurious. The gRPC stream will be closed and no further
+  // messages will be sent to the external processing server during the lifetime of the filter.
+  // This is to protect Envoy from a malformed server.
```

Comment:
HCM already knows what max_request_headers_kb_ should be, so it seems dangerous to allow the operator to configure this. We should  verify that whatever the ext_proc server returns maintains the invariant that total headers are less than the HCM threshold I think. CC @yanavlasov

Reply:
Thanks for the comments! 
Can we check against the header size limit config in HCM in ext_proc filter context?   And if we do, is that the intention here? i.e, do we really want to say the response headers size from ext_proc server should not exceed 60kib (hcm default), or should it be a much smaller number, thus the combined header after mutation won't be too much?  For example, if the ext_proc server response header mutation are all append headers, 60kib, then plus the original 60kib from downstream client, will ends up 120kib headers.

Reply:
We should have the combined header size respect the HCM limit. We don't want to have ext_proc act as a backdoor to violate header size limits that bound memory overheads of in-flight requests. In your example, the combined size should be   under 60 kb.

Reply:
I see. So we need to add a check here to make sure the headers.size is under 60kb

https://github.com/envoyproxy/envoy/blob/c6292584f59eac45bf43f3dba7b0f8c36b3ab0bc/source/extensions/filters/http/ext_proc/mutation_utils.cc#L116

However, do we need to enforce a check for the response size here:

https://github.com/envoyproxy/envoy/blob/c6292584f59eac45bf43f3dba7b0f8c36b3ab0bc/api/envoy/service/ext_proc/v3/external_processor.proto#L126

or just leave it open?

If we leave it open, and if a ext_proc server sent back a huge response, which contains huge amount of header mutations, like in tens of thousands,  it may cause CPU hog for Envoy when it is calling applyHeaderMutations.

Reply:
It's a good question. Ideally we can bound the size of the response proto as well. I'm not sure how practical this is, it probably depends on Envoy and Google gRPC implementations. I don't think we can easily do it from ext_proc alone can we?

Reply:
I don't think it's an ext_proc specific problem.
In other words, what's the difference between this and a byzantine upstream server that sends a response with a large header? AFAICT the same guards should be available (not only for headers size limits, but also for other types of attack vectors).

Reply:
If there is such size check for upstream response, please point me the code.  
I checked ext_authz, there is no response size check, but its external service might be trusted. It's security posture does not mention robust to the side stream.

Overall, can we do these:

1) The response proto size < 10Mb
2) if the response is headers or trailers(for either downstream request or upstream response ), the number of headers it contains < 100. The size of the headers < 60Kb.  And at the end after applyHeaderMutation, the end result header size < 60Kb.
3) If the response is body, no further size check other than 1).

Reply:
Sorry, my last reply was related to the actual response from the side-stream (the caps are being enforced [here](https://github.com/envoyproxy/envoy/blob/cab680b0ef801e269f9a50cd745728d4ae8b86e6/source/common/http/http2/codec_impl.cc#L1485)).

For the Ext-Proc response fields, there should be additional validation.

Reply:
The big question is how to surface HCM limits on the header map to the ext_proc filter. These can be added to the StreamInfo or filter state under well known key.

The number of header additions would be limited by the max number of headers. Number of removals I think should be limited too, although it is unclear what this limit should be. Maybe as a first step limit it by them max number of headers as well.

Reply:
@adisuissa IIUC, the caps enforced in the codec_impl is for the actual header of the messages from ext_proc server to Envoy. In our case, the headers encoded in ProcessingResponse: https://github.com/envoyproxy/envoy/blob/0ae464feb4b3d1655baed1087cc97339fc644712/api/envoy/service/ext_proc/v3/external_processor.proto#L126 
is actually the data. Thus the code pointed above won't help. Please correct me if I am wrong.

Reply:
@yanavlasov Thanks for the comments! So, here is the summary:
1) let's try surface the HCM limits in the streamInfo.
2) Enforce both the header additions and header removals in the response against the HCM max number of headers (100 by default) and max headers size(60kb).
3) The ends up headers size and head number after applyHeaderMutation() should also be checked against the HCM limits.

One more question:
Do we need to enforce some limit for the ProcessingResponse body size?
https://github.com/envoyproxy/envoy/blob/0ae464feb4b3d1655baed1087cc97339fc644712/api/envoy/service/ext_proc/v3/external_processor.proto#L215

If we do, what's the size limit for it?

Reply:
There should be a limit. I'm unsure what this should be though. One approach is to limit the size the gRPC messages through the gRPC channel options and allow changing this limit in the ext_proc config.

Reply:
Based on an offline discussion with @stevenzzzz , he recommends no need to enforce limit on the body size, since it will be handled by the buffer management with below code path:

de/encoding_state_.handleBodyResponse(response->response_body()); 
https://github.com/envoyproxy/envoy/blob/2ab9d6b7f816acbbde3cecc9357efc4f0960c9c8/source/extensions/filters/http/ext_proc/processor_state.cc#L208

tries to buffer the returned body in the stream buffer, where it eventually get to this:

https://github.com/envoyproxy/envoy/blob/2ab9d6b7f816acbbde3cecc9357efc4f0960c9c8/source/common/buffer/watermark_buffer.cc#L29

if it's too large, it would trigger the flow control mechanism and cause a 500 response too large on response path, and a 400 error on the response path.